datafusion/common/src/error.rs

```diff
@@ -36,6 +36,9 @@ use sqlparser::parser::ParserError;
 /// Result type for operations that could result in an [DataFusionError]
 pub type Result<T> = result::Result<T, DataFusionError>;
 
+/// Result type for operations that could result in an [DataFusionError] and needs to be shared (wrapped into `Arc`).
+pub type SharedResult<T> = result::Result<T, Arc<DataFusionError>>;
```

Comment:
Given this PR never actually returns this type, do we need to declare it here?

Reply:
Well, I observed that `Result<T, Arc<Error>` is quite common approach to allow clonning errors in datafustion & arrow.

And not to miss such kind of errors, I think it would be nice to have it as general type. 
Take a look at #4992 and even current `find_root` impl, it's easy to forget that error could be wrapped into `Arc` actually (IMO)

I believe this type should be used when it's needed to explicitly declare that error is wrapped into `Arc`. And could be used in other places later.

Reply:
> allow clonning errors in datafustion & arrow.

I don't see any other examples of this pattern?

Reply:
See https://github.com/apache/arrow-datafusion/pull/4992 for the code (that does indeed walk over Arcs). It was non ideal however. Figuring out some way to avoid Arc'ing errors would be better in my opinion (not for this PR)